### PR TITLE
Feature/27 Add read entry endpoint

### PR DESCRIPTION
**Pull request by Rumpelshtinskiy**
Closes #

### Please check if the PR fulfills these requirements

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] CHANGELOG.md has been updated (for bug fixes / features / docs)

### What was changed?

Added an endpoint GET /api/v1/b/:bucket_name/:entry_name to get a record of an entry

### CODE REVIEW

File: src/main/java/org/reduct/client/ReductClient.java [hunk 1]
```diff
@@ -17,15 +17,20 @@
  * Base class for all clients.
  */
 public abstract class ReductClient {
+   private static final String ERROR_400 = "Posted content bigger or smaller than content-length";
```
Comment (atimin) [inline src/main/java/org/reduct/client/ReductClient.java:20]: The server sends text messages with the error descriptions. We don't need re-implement this.

File: src/main/java/org/reduct/client/ReductClient.java [hunk 2]
```diff
@@ -54,4 +59,29 @@ <T> HttpResponse<String> sendRequest(String url, Method method) throws ReductExc
...
    }
+
+   public <T> HttpResponse<T> send(HttpRequest.Builder builder, HttpResponse.BodyHandler<T> bodyHandler) {
+      try {
+         HttpResponse<T> httpResponse = getHttpClient().send(builder.build(), bodyHandler);
+         switch (httpResponse.statusCode()) {
+            case 200 -> {
+               return httpResponse;
+            }
+            case 400 -> throw new ReductException(ERROR_400, httpResponse.statusCode());
```
Comment (atimin) [inline src/main/java/org/reduct/client/ReductClient.java:70]: Just use the server's message from `x-reduct-error` header.

File: src/main/java/org/reduct/model/entry/Entry.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+import java.io.Serializable;
+
+@Getter
+@Setter
+@AllArgsConstructor
+@Builder
+public class Entry<BODY_CLASS extends Serializable> {
+    private String name;
+    private Long timestamp;
+    private final BODY_CLASS body;
```
Comment (atimin) [inline src/main/java/org/reduct/model/entry/Entry.java:18]: There is some meta information. Content type and labels.

File: src/main/java/org/reduct/client/ReductEntryClient.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                .POST(HttpRequest.BodyPublishers.ofByteArray(entry.getByteBodyArray()));
+
+        if(isNotBlank(getToken())) {
+            builder.headers("Authorization", "Bearer %s".formatted(getToken()));
+        }
+        return send(builder, HttpResponse.BodyHandlers.ofString()).body();
+    }
+
+    @Override
+    public <T extends Serializable> Entry<T> getRecord(Bucket bucket, Entry<?> entry) throws ReductException, ReductSDKException, IllegalArgumentException {
```
Comment (atimin) [inline src/main/java/org/reduct/client/ReductEntryClient.java:68]: The method has the `getRecord ` name but returns Entry. To get a record, we need only an entry name. Actually, we don't need  the Entry class, it should be Record. Could it be possible to use the same approach as we have in other SDK?:

```
Bucket bucket = client.getBucket("bucket_name");
Record rec = bucket.read("entry_name", timestamp);

```

File: src/main/java/org/reduct/client/ReductEntryClient.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        this.serverProperties = serverProperties;
+        this.token = accessToken;
+    }
+
+    public ReductEntryClient(ServerProperties serverProperties, String accessToken) {
+        this(serverProperties, HttpClient.newHttpClient(), accessToken);
+    }
+
+    @Override
+    public String writeRecord(@NonNull Bucket bucket, @NonNull Entry<?> entry) throws ReductException, ReductSDKException, IllegalArgumentException {
```
Comment (atimin) [inline src/main/java/org/reduct/client/ReductEntryClient.java:44]:
```
Bucket bucket = client.getBucket("bucket");
bucket.writeRecord("entry_name", record);
```

**Review by atimin — CHANGES_REQUESTED** [review]
How do you plan to read and write large blobs in chunks?

File: src/main/java/org/reduct/model/entry/Entry.java (hunk 3 above)
Comment (Rumpelshtinskiy) [reply src/main/java/org/reduct/model/entry/Entry.java:18]: Not yet, not in this branch

File: src/main/java/org/reduct/client/ReductClient.java (hunk 2 above)
Comment (Rumpelshtinskiy) [reply src/main/java/org/reduct/client/ReductClient.java:70]: Fixed

File: src/main/java/org/reduct/client/ReductClient.java [hunk 6]
```diff
@@ -54,4 +58,110 @@ <T> HttpResponse<String> sendRequest(String url, Method method) throws ReductExc
          throw new ReductSDKException("Thread has been interrupted while processing the request", e);
       }
    }
+
+   public Bucket getBucket(String name) throws ReductException, ReductSDKException, IllegalArgumentException {
+      return new Bucket(name, this).read();
+   }
+
+   public Bucket createBucket(String name) throws ReductException, ReductSDKException, IllegalArgumentException {
```
Comment (atimin) [inline src/main/java/org/reduct/client/ReductClient.java:66]: Let's pass settings during the creation of the bucket:

1.  We need BucketSettings class for serialization
2. Add Bukcet.setSettings and Bucket.getSettings methods to mange the settings.

File: src/main/java/org/reduct/model/bucket/Bucket.java [hunk 7]
```diff
@@ -37,11 +66,278 @@ public class Bucket {
...
+    /**
+     * Get a record from an entry.
+     * @param entry
+     * @return
+     * @param <T>
+     * @throws ReductException
+     * @throws ReductSDKException
+     * @throws IllegalArgumentException
+     */
+    public <T extends Serializable> Entry<T> getRecord(Entry<?> entry) throws ReductException, ReductSDKException, IllegalArgumentException {
```
Comment (atimin) [inline src/main/java/org/reduct/model/bucket/Bucket.java:200]:
```
Record readRecord(String entry_name, DateTime time)
```

File: src/main/java/org/reduct/model/bucket/Bucket.java [hunk 8]
```diff
@@ -37,11 +66,278 @@ public class Bucket {
...
+    /**
+     * Get only meta information about record.
+     * @param entry
+     * @return
+     * @param <T>
+     * @throws ReductException
+     * @throws ReductSDKException
+     * @throws IllegalArgumentException
+     */
+    public <T extends Serializable> Entry<T> getMetaInfo(Entry<?> entry) throws ReductException, ReductSDKException, IllegalArgumentException {
```
Comment (atimin) [inline src/main/java/org/reduct/model/bucket/Bucket.java:229]:
```
Record getMetaInfo(String entryName, DateTime time)
```

File: src/main/java/org/reduct/model/bucket/Bucket.java [hunk 9]
```diff
@@ -37,11 +66,278 @@ public class Bucket {
...
+     *
+     * The method responds with a JSON document containing an ID which should be used to read records with the following endpoint:
+     * GET /b/:bucket_name/:entry_name?q=ID.
+     * @param entry
+     * @param start
+     * @param stop
+     * @param ttl
+     * @return
+     */
+    public Query query(Entry<?> entry, Integer start, Integer stop, Integer ttl) {
```
Comment (atimin) [inline src/main/java/org/reduct/model/bucket/Bucket.java:324]:
```
public Iterator<Record> query(Entry<?> entry, Integer start, Integer stop, Integer ttl)
````

File: src/main/java/org/reduct/model/bucket/Bucket.java [hunk 10]
```diff
@@ -37,11 +66,278 @@ public class Bucket {
...
+    }
+
+    /**
+     * Write a record to an entry.
+     * @param entry
+     * @throws ReductException
+     * @throws ReductSDKException
+     * @throws IllegalArgumentException
+     */
+    public void writeRecord(@NonNull Entry<?> entry) throws ReductException, ReductSDKException, IllegalArgumentException {
```
Comment (atimin) [inline src/main/java/org/reduct/model/bucket/Bucket.java:146]:
```
public void writeRecord(Record record)
```

for chunks

```
public void writeRecord(Record record, Iterator<Record> stream)
```

File: src/main/java/org/reduct/client/ReductClient.java (hunk 6 above)
Comment (Rumpelshtinskiy) [reply src/main/java/org/reduct/client/ReductClient.java:66]: Completed

File: src/main/java/org/reduct/model/bucket/Bucket.java (hunk 7 above)
Comment (Rumpelshtinskiy) [reply src/main/java/org/reduct/model/bucket/Bucket.java:200]: Completed

File: src/main/java/org/reduct/model/bucket/Bucket.java (hunk 8 above)
Comment (Rumpelshtinskiy) [reply src/main/java/org/reduct/model/bucket/Bucket.java:229]: Completed

File: src/main/java/org/reduct/model/bucket/Bucket.java (hunk 9 above)
Comment (Rumpelshtinskiy) [reply src/main/java/org/reduct/model/bucket/Bucket.java:324]: Completed

File: src/main/java/org/reduct/client/ReductClient.java (hunk 1 above)
Comment (Rumpelshtinskiy) [reply src/main/java/org/reduct/client/ReductClient.java]: done

File: src/main/java/org/reduct/client/ReductEntryClient.java (hunk 5 above)
Comment (Rumpelshtinskiy) [reply src/main/java/org/reduct/client/ReductEntryClient.java]: done

File: src/main/java/org/reduct/utils/Deserializer.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+import lombok.experimental.UtilityClass;
+
+import java.io.BufferedInputStream;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+@UtilityClass
+public class Deserializer {
```
[the remark this answers is not in the document]
Comment (Rumpelshtinskiy) [reply src/main/java/org/reduct/utils/Deserializer.java]: deleted

File: src/main/java/org/reduct/utils/Serializer.java [hunk 12]
```diff
@@ -0,0 +1,41 @@
+package org.reduct.utils;
+
+
+import lombok.experimental.UtilityClass;
+
+import java.nio.ByteBuffer;
+
+@UtilityClass
+public class Serializer {
```
[the remark this answers is not in the document]
Comment (Rumpelshtinskiy) [reply src/main/java/org/reduct/utils/Serializer.java]: deleted

File: src/main/java/org/reduct/model/bucket/Bucket.java (hunk 10 above)
Comment (Rumpelshtinskiy) [reply src/main/java/org/reduct/model/bucket/Bucket.java]: Done

**Pull request merged**